# Morton Code vs Hash map - Conta apenas vizinhos e Normalizado

#### No buffer debug foi gravada a distancia que é percorrida ao calcular a densidade, correndo todos os vizinhos
#### Assim, se para a particula i (no indice 1) vamos vizitar as particulas A B e C e estas estão nos indices 3,4,5
#### O buffer debug, vai guardar para a particula i uma distancia de 3+4+5 = 12 (no caso de A B e C serem vizinhos)

#### Este processo é feito por 500 frames e as distancias são todas somadas.
#### Assim, vai se fazer uma média da distancia percorrida por particula nesses 500 frames.

## Quando digo normalizado, quere a cada iteração é guardada uma distancia média e não a soma das distancias.
## Ou seja, para cada frame é calculada a distancia, em média, de uma particula com os seus vizinhos e esse valor é que é somado todos os frames

In [1]:
import csv
import pandas as pd

In [2]:
hashMap = pd.read_csv('data/HashMap_500Frames_Vizinhos_normal.txt',sep=' ',usecols= ['SUM','FRAMES'])
mortonCode = pd.read_csv('data/MortonCode_500Frames_Vizinhos_normal.txt',sep=' ',usecols= ['SUM','FRAMES'])

In [3]:
hashMap

SUM  FRAMES
0        354600000.0     500
1        363145000.0     500
2        366933000.0     500
3        359046000.0     500
4        357867000.0     500
...              ...     ...
1727995  323221000.0     500
1727996  321161000.0     500
1727997  329770000.0     500
1727998  325791000.0     500
1727999  337110000.0     500

[1728000 rows x 2 columns]

In [4]:
mortonCode

SUM  FRAMES
0         60065.7     500
1         43897.3     500
2         46402.4     500
3         40829.7     500
4         41649.7     500
...           ...     ...
1727995  124756.0     500
1727996  147720.0     500
1727997  111146.0     500
1727998  112428.0     500
1727999   97619.0     500

[1728000 rows x 2 columns]

In [5]:
type(mortonCode['SUM'][0])
type(mortonCode['FRAMES'][0])

numpy.int64

In [6]:
#Vai criar um array qe vai ter como resultado Mortoncode[SUM][i]/Mortoncode[FRAMES][i]
mortonCodeDiv=mortonCode['SUM']/mortonCode['FRAMES']
hashMapDiv=hashMap['SUM']/hashMap['FRAMES']

In [7]:
#Soma todos os resultados e divide pelo numero de particulas, assim temos a média por particula
mortonCodeResult = mortonCodeDiv.sum()/mortonCodeDiv.size
HashMapResult = hashMapDiv.sum()/hashMapDiv.size

In [8]:
print("Média da soma de distancia percurrida por cada particula ao comparar cada particula na densidade")
print("Quando a particula i vai para calcular a sua densidade, esta vai comparar com X número de outras particulas")
print("O resultado é a soma das distancias, distancia de indices dentro do buffer, de todas as particulas X que são vizinhas")
print("Exemplo: a particula A esta no indice 2 e tem 2 possiveis vizinhos B e C. B esta no indice 3 e é vizinho e C no indice 4 e não é vizinho. Assim o resultado seria 1 de distancia total")
print("Resultado Morton Code:",mortonCodeResult)
print("Resultado Hash Map:",HashMapResult)

Média da soma de distancia percurrida por cada particula ao comparar cada particula na densidade
Quando a particula i vai para calcular a sua densidade, esta vai comparar com X número de outras particulas
O resultado é a soma das distancias, distancia de indices dentro do buffer, de todas as particulas X que são vizinhas
Exemplo: a particula A esta no indice 2 e tem 2 possiveis vizinhos B e C. B esta no indice 3 e é vizinho e C no indice 4 e não é vizinho. Assim o resultado seria 1 de distancia total
Resultado Morton Code: 7108.3918543992995
Resultado Hash Map: 396030.925162037


# Assim podemos ver que no morton code, uma particula está, em média a 7109 indices de distancia da sua particula vizinha.
# Em comparação com o Hash Map, temos uma distancia média de 396031 indices
## Morton code tem uma média cerca de 55 vezes menor

## Com isto podemos concluir que o morton code é muito mais provavel de tirar partido da cache da placa gráfica

## Hash Map density 77 ms
## Morton Code density 7 ms -> 11 vezes mais rápido